File: ddd-briscola-web/src/main/ts/model/game.ts
module Model {

  import Option = Std.Option

  export enum GameStateKind {
    empty, active, dropped, finished
  }

  export enum BriscolaEventKind {
    gameStarted, cardPlayed, gameDropped
  }

  export enum Seed {
    bastoni, coppe, denari, spade
  }
  
  export enum DropReasonKind {
    playerLeft  
  }

  export type BriscolaEvent = CardPlayed | GameStarted | GameDropped 
  
  export class CardPlayed extends DomainEvent {
    constructor(public game: GameState, public player: Player, public card: Card) {
      super(BriscolaEventKind[BriscolaEventKind.cardPlayed])
    }
  }

  export class GameStarted extends DomainEvent {
    constructor(public game: ActiveGameState) {
      super(BriscolaEventKind[BriscolaEventKind.gameStarted])
    }
  }
  
  export class GameDropped extends DomainEvent {
    constructor(public game: GameState, dropReason:DropReason) {
      super(BriscolaEventKind[BriscolaEventKind.gameDropped])
    }
  }

  export type GameState = FinalGameState | ActiveGameState | DroppedGameState

  export class FinalGameState extends Aggregate {
    constructor(
      public self: Ws.Path,
      public briscolaCard: Card,
      public playersOrderByPoints: PlayerFinalState[],
      public winner: PlayerFinalState) {
      super()
    }
  }

  export class ActiveGameState extends Aggregate {
    constructor(
      public self: Ws.Path,
      public briscolaCard: Card,
      public moves: Move[],
      public nextPlayers: Player[],
      public currentPlayer: Player,
      public isLastHandTurn: Boolean,
      public isLastGameTurn: Boolean,
      public players: Player[],
      public playerState: Option<PlayerState>,
      public deckCardsNumber: number) {
      super()
    }
  }
  
  export class DropReason {
  }
  
  export class PlayerLeft extends DropReason {
    constructor(public player:Player, reason:Option<string>) {
      super();
    }
  }
  
  export class DroppedGameState extends Aggregate {
    constructor(
      public self: Ws.Path,
      public briscolaCard: Card,
      public moves: Move[],
      public nextPlayers: Player[],
      public dropReason:DropReason) {
      super()
    }
  }


  export interface Card {
    seed: Seed
    "number": number
  }

  export interface Move {
    player: Player
    card: Card
  }

  export interface PlayerFinalState {
    player: Player
    points: number
    score: Card[]
  }

  export interface PlayerState {
    self: Ws.Path
    player: Player
    cards: Card[]
    score: Card[]
  }


  export module BriscolaEvent {
    export function fold<T>(
      gameStarted: (p: GameStarted) => T, cardPlayed: (p: CardPlayed) => T): (p: BriscolaEvent) => T {
      return p => {
        if (p instanceof GameStarted) return gameStarted(p)
        else if (p instanceof CardPlayed) return cardPlayed(p)
        else {
          console.log("unrecognized BriscolaEvent")
          console.log(p)
          return Util.fail<T>("unrecognized BriscolaEvent ")
        }
      }
    }
  }

  export module GameState {
    export function fold<T>(
      activeGameState: (p:ActiveGameState) => T, 
      finalGameState: (p:FinalGameState) => T,
      droppedGameState: (p:DroppedGameState) => T ): (p: GameState) => T {
      return p => {
        if (p instanceof ActiveGameState) return activeGameState(p)
        else if (p instanceof FinalGameState) return finalGameState(p)
        else if (p instanceof DroppedGameState) return droppedGameState(p)
        else {
          console.log("unrecognized GameState")
          console.log(p)
          return Util.fail<T>("unrecognized GameState ")
        }
      }
    }
  }

}

File: ddd-briscola/src/main/scala/com/github/gdefacci/briscola/game/game.scala
package com.github.gdefacci.briscola.game

import com.github.gdefacci.ddd._

import scalaz.{-\/, \/, \/-}

import com.github.gdefacci.briscola.player._

trait GameDecider extends Decider[GameState, BriscolaCommand, BriscolaEvent, BriscolaError] {
  
  def nextId:GameId
  def playerById(name:PlayerId):Option[Player]
  
  def apply(s:GameState, cmd:BriscolaCommand):BriscolaError \/ Seq[BriscolaEvent] = {
    (s, cmd) match {
      case (EmptyGameState, StartGame(gamePlayers)) =>

        GamePlayersValidator.withValidPlayersAndTeams(gamePlayers, playerById(_)) { (players, teams) =>
          val (deck, plyrs) = players.foldLeft(Deck.initial -> Seq.empty[PlayerState]) { (acc, player) =>
            val (deck, currPlayers) = acc
            val (cards, newDeck) = deck.takeCards(3)
            newDeck -> (currPlayers ++ Seq(PlayerState(player.id, cards, Score.empty)))
          }

          Seq(GameStarted(ActiveGameState(nextId, deck.briscolaCard(players.size), deck, Nil, plyrs, teams)))
        }
        
      case (EmptyGameState, _) => 
        -\/(GameNotStarted)
        
      case (gm:ActiveGameState, PlayCard(pid, card)) if gm.currentPlayer.id == pid && gm.currentPlayer.cards.contains(card) => 
        \/-(Seq(CardPlayed(pid, card)))
      
      case (gm:ActiveGameState, PlayCard(pid, card)) if gm.currentPlayer.id == pid && !gm.currentPlayer.cards.contains(card) => 
        -\/(PlayerDoesNotOwnCard(pid, card, gm.currentPlayer.cards))
      
      case (gm:ActiveGameState, PlayCard(pid, _)) if !(gm.players.map(_.id) contains pid) => 
        -\/(InvalidPlayer(pid))
      
      case (gm:ActiveGameState, PlayCard(pid, _)) if gm.currentPlayer.id != pid => 
        -\/(InvalidTurn(pid, gm.currentPlayer.id))
      
      case (_:ActiveGameState, StartGame(players)) => 
        -\/(GameAlreadyStarted)
      
      case (_:ActiveGameState, PlayerDropGame(player, reason)) => 
        \/-(Seq(GameDropped(PlayerLeft(player, reason))))
      
      case (_:DroppedGameState, _) => 
        -\/(GameAlreadyDropped)
      
      case (_:FinalGameState, _) => 
        -\/(GameAlreadyFinished)
      
    }
  }
  
}

trait GameEvolver extends Evolver[GameState, BriscolaEvent] {
 
  def playHand(moves:Seq[Move], gameSeed:Seed, deck:Deck):(Deck, Seq[PlayerState]) = {
    val movesWithGameSeed = moves.filter(_.card.seed == gameSeed)
    val winner = if (movesWithGameSeed.nonEmpty) {
      movesWithGameSeed.sortBy(_.card.points).last.player.id
    } else {
      val firstCardSeed = moves.head.card.seed
      moves.filter(_.card.seed == firstCardSeed).sortBy(_.card.points).last.player.id
    }
    val winnerCards = moves.map(_.card)
    val indexOfWinner = moves.indexWhere(_.player.id == winner)
    val len = moves.length
    val idxStates = moves.zipWithIndex.map { p =>
      val (mv, idx) = p
      val newIndex = if (idx < indexOfWinner) len + idx - indexOfWinner else idx - indexOfWinner
      val wonCards = if (idx == indexOfWinner) winnerCards else Nil
      newIndex -> PlayerState(mv.player.id, mv.player.cards.filter(_!=mv.card), mv.player.score.add(wonCards)) 
    }
    idxStates.sortBy(_._1).map(_._2).foldLeft(deck -> Seq.empty[PlayerState]) { (acc, playerState) =>
      val (deck, states) = acc
      val (card, newDeck) = deck.takeCards(1)
      newDeck -> (states ++ Seq(PlayerState(playerState.id, playerState.cards ++ card, playerState.score)))
    }
  }

  def apply(s:GameState, event:BriscolaEvent):GameState = {
    (s,event) match {
      case (EmptyGameState, GameStarted(state)) => 
        state
        
      case (gm @ ActiveGameState(id, briscolaCard, deck, moves, nextPlayers, team), CardPlayed(cardPlayerId, card)) =>
        val newMoves = gm.moves ++ Seq(Move(gm.currentPlayer.copy(cards = gm.currentPlayer.cards.filter(_ != card)), card))
        if (gm.isLastHandTurn) {
          
          val (newDeck, newPlayersState) = playHand(newMoves, briscolaCard.seed, gm.deck)
          
          if (gm.isLastGameTurn) {
            val nsts = newPlayersState.map(s => PlayerFinalState(s.id, s.score ))
            FinalGameState(gm.id, gm.briscolaCard, nsts, team) 
          } else {
            gm.copy(deck = newDeck, moves = Nil, nextPlayers = newPlayersState)
          }
          
        } else {
          gm.copy(moves = newMoves, nextPlayers = gm.nextPlayers.tail)
        }
        
      case (ActiveGameState(id, briscolaCard, deck, moves, nextPlayers, team), GameDropped(reason))  =>
        DroppedGameState(id, briscolaCard, deck, moves, nextPlayers, reason, team)
        
      case (s, cmd) => 
        throw new RuntimeException(s"forbidden condition state:$s event:$event")
    }
  }
}